Feature reimbursement forms

KaiKaiai opened the pull request.

khangluong2004 committed:
Prepare route, button for the form. Will fill form tmr

Troppydash committed:
Added Submission API

khangluong2004 committed:
Add template and authen for forms

khangluong2004 committed:
Finish the form

khangluong2004 committed:
Finish integrating. I'm stealing kevin's id

Troppydash committed:
added reimbursement card, connected reimbursement apis

khangluong2004 committed:
Turn the "js-style" ts back to ts :0

khangluong2004 committed:
Pdf upload

khangluong2004 committed:
Push upload file

Troppydash committed:
fixed setting token in the user database

Troppydash committed:
added treasurer specific dashboard logic

khangluong2004 committed:
Add department and finish test upload

Troppydash committed:
Added department field, updated reimbursement render

thebowenfeng commented on src/auth/google.ts:
```diff
@@ -17,31 +17,28 @@ provider.addScope("https://www.googleapis.com/auth/drive.file");
+export async function googleSignIn(): Promise<User> {
+  const result: UserCredential = await signInWithPopup(auth, provider);
+  const credential = GoogleAuthProvider.credentialFromResult(result);
+  console.log(credential);
```
Prolly shouldn't be logging this lol

thebowenfeng commented on src/auth/types.ts:
```diff
@@ -4,8 +4,34 @@ export interface User {
+export function createUser(
+  obj: any
+): User {
+  if (obj == null){
```
Might not be a big issue depending on how your logic is setup but shallow equality here means undefined obj also passes the check (i.e when obj is undefined).

Troppydash replied:
Probably keep it as shallow compare; we need undefined to be truthy else the destructuring below fails.

thebowenfeng commented on src/database/reimbursement.ts:
```diff
@@ -0,0 +1,89 @@
+
+const db = getFirestore(app);
+
+export type DepartmentEnum = "IT" | "Events" | "Competition" | "Education" | "Industry" | "Project" | "Diversity" | "Publicity" | "Product";
```
Easier if you define an actual enum and the can just access the property like Department.Publicity

khangluong2004 replied:
I kept it as a group of strings so it's easier to use those strings as selection options directly though

thebowenfeng commented on src/database/reimbursement.ts:
```diff
@@ -0,0 +1,89 @@
+
+  // store the document
+  const ref = doc(collection(db, "reimbursement"));
+  await setDoc(ref, submission);
```
Should prolly handle errors at this point in case setDoc fails.

thebowenfeng commented on src/database/reimbursement.ts:
```diff
@@ -0,0 +1,89 @@
+  );
+
+  const snapshot = await getDocs(q);
+  let result: Reimbursement[] = [];
```
Can be replaced by map

khangluong2004 committed:
Add pre-filled mail

Troppydash committed:
updated website name in index.html

Troppydash committed:
auth quality of life

1. added automatic sign in
2. added logout button and logout
3. added analytic page skeleton
4. changed 404 page to home

Troppydash committed:
Added analytics

khangluong2004 committed:
Resolve Bowen's comment

khangluong2004 committed:
Flip to map and leave the null thing for now
